src/core/cdr/include/dds/cdr/dds_cdrstream.h

```diff
@@ -44,9 +71,58 @@ typedef struct dds_ostreamLE {
   dds_ostream_t x;
 } dds_ostreamLE_t;
 
+typedef struct dds_cdrstream_desc_key {
+  uint32_t ops_offs;   /* Offset for key ops */
+  uint32_t idx;        /* Key index (used for key order) */
+} dds_cdrstream_desc_key_t;
+
+typedef struct dds_cdrstream_desc_key_seq {
+  uint32_t nkeys;
+  dds_cdrstream_desc_key_t *keys;
+} dds_cdrstream_desc_key_seq_t;
+
+typedef struct dds_cdrstream_desc_op_seq {
+  uint32_t nops;    /* Number of words in ops (which >= number of ops stored in preproc output) */
+  uint32_t *ops;    /* Marshalling meta data */
+} dds_cdrstream_desc_op_seq_t;
+
+struct dds_cdrstream_desc {
+  uint32_t size;    /* Size of type */
+  uint32_t align;   /* Alignment of top-level type */
+  uint32_t flagset; /* Flags */
+  dds_cdrstream_desc_key_seq_t keys;
+  dds_cdrstream_desc_op_seq_t ops;
+  size_t opt_size_xcdr1;
+  size_t opt_size_xcdr2;
+};
+
+#ifndef NDEBUG
+typedef struct align { uint32_t a; } align_t;
+#define ALIGN(n) ((n).a)
+#else
+typedef uint32_t align_t;
+#define ALIGN(n) (n)
+#endif
+
 DDSRT_STATIC_ASSERT (offsetof (dds_ostreamLE_t, x) == 0);
+
 DDSRT_STATIC_ASSERT (offsetof (dds_ostreamBE_t, x) == 0);
 
+inline align_t dds_cdr_get_align (uint32_t xcdr_version, uint32_t size)
+{
+#ifndef NDEBUG
+#define MK_ALIGN(n) (struct align){(n)}
+#else
+#define MK_ALIGN(n) (n)
+#endif
+  if (size > 4)
+    return xcdr_version == DDS_CDR_ENC_VERSION_2 ? MK_ALIGN(4) : MK_ALIGN(8);
+  return MK_ALIGN(size);
+#undef MK_ALIGN
+}
+
+uint32_t dds_cdr_alignto_clear_and_resize (dds_ostream_t * __restrict s, align_t a, uint32_t extra);
```

Comment:
I wonder if it doesn't make more sense to split this function into an internal one taking `align_t` and an externally visible wrapper taking a `uint32_t` and calling the internal one. That way `align_t` and `dds_cdr_get_align` don't have to be exported.

For that matter, the only two external call sites (in this repo, haven't checked python and C++ yet) only call it with `dds_cdr_get_align (s->m_xcdr_version, 4)` to pad it out. Arguably, that means the externally visible function could simply be `dds_cdr_alignto4_clear_and_resize`.

Reply:
That's a good point, I've refactored this in [09fee43](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1468/commits/09fee435c23d47d12ad16a73bcc97688cf9dc15d)